### ISSUE TITLE
Sifnode Peggy2 Load Testing: Creating an increasing number of pools

### ISSUE THREAD

**Issue by pandaring2you** [Sifchain/sifnode#3020]
When creating an increasing number of pools - check relative performance (is it linearly increasing?)

**Comment by jzvikart** [comment]
Scenario description: https://www.notion.so/sifchain/Rewards-2-0-Load-Testing-972fbe73b04440cd87232aa60a3146c5

Open questions:
- How do we measure block time?
- Can we change block time?
- If the calculations fit into existing block time, does the time of calculations influence block time? (i.e. is the block mined on a timer with a constant firing rate)

**Comment by jzvikart** [comment]
@timlind said about changing block time:
> its part of consensus config in genesis.json
> timeout_commit + timeout_precommit

**Comment by jzvikart** [comment]
@joneskm's script for querying unique and total LPs:

```
#!/bin/bash

set -eu

NODE="https://rpc-archive.sifchain.finance:443"

rm -f all-lps.txt

for i in {0..61} #61 manually checked - the 62nd is empty!
do
   echo "Processing page $i"
   sifnoded q clp all-lp --limit=200 --height 7759160 --node $NODE --output json --page $i | jq -r '.liquidity_providers | .[] | .liquidity_provider_address' >> all-lps.txt
done

cat all-lps.txt | sort | uniq > unique-lps.txt

wc -l all-lps.txt
wc -l unique-lps.txt
```

**Comment by jzvikart** [comment]
@joneskm If we wanted to "apply" the numbers from production to the test, we would have to run the test with these parameters:
- `--number-of-wallets 7025`
- `--number-of-pools 100` (== number_of_tokens)
- `--liquidity-providers-per-wallet 2` (== number of tokens per wallet)

Thus, each of 7025 wallets would provide liquidity to 2 pools (chosen randomly out of 100 total pools), so we would have 7025 unique and 7025 * 2 = 14050 total liquidity providers.

The test implicitly assumes:
- we're only dealing with symmetrical liquidity (i.e. ` native_amount == external amount`)
- total number of different tokens == total number of liquidity pools
- number of different tokens per address == number of liquidity providers per address

**Comment by jzvikart** [comment]
@daechoi mentioned that one of main concerns is also the gossip time. We should try to run this test with multiple nodes.

Can anybody show me how to set up multiple nodes via CLI?

**Comment by pandaring2you** [comment]
MVP eta tomorrow.

**Comment by jzvikart** [comment]
Iteration 1:
- number_of_liuquidity_pools: 100
- number_of_wallets: 8000
- liquidity_providers_per_wallet: 3
- test_duration_blocks 100

Creating 8000 accounts via `add-genesis-account` turns out to be slow, since each invocation has to load, parse, modify and save the genesis file. The time needed for that increases with every added account.

**Comment by jzvikart** [comment]
Test interation 2:
- setup of accounts now works (implemented via directly modifying genesis.json)
- `sifnoded query clp lplist` crashes the test because the number of results exceed page size

Next step:
- implement reading of results with pagination

**Comment by jzvikart** [comment]
### Test iteration 3:
- Setup same as above
- Implemented paging
- Current result: test failing because of intermittent HTTP 500 errors for http://.../block_results
- Next steps:
  - implement retry logic + leave sifnoded process running so that we can debug
  - investigate with Brian, James and Dae what is happening

**Comment by jzvikart** [comment]
### Test iteration 4:
- Added some general debugging/troubleshooting logic
- Dae/Brian: try first with just 5 pools, if it still fails also reduce number of wallets to 1000

**Comment by jzvikart** [comment]
### Test iteration #5

Setup: 5 pools / 8000 wallets / 3 liquidity providers per wallet, single node running on localhost

Results:
- HTTP 500 for `block_results` occures 2001 times, >90% failing when LPPD is active, OK otherwise
- Average block time over 100 blocks:
  - Neither: 5.03 s/block
  - Rewards only: 5.03 s/block
  - Rewards + LPPD: 13.39 s/block
  - LPPD only: 14.63 s/block

**Comment by jzvikart** [comment]
### Test iteration 6:

Scenario: 100 pools / 8000 wallets / 3 liquidity providers per wallet, single node running on localhost
Branch: fix/lppd_rewards_block_time (63830c223c520a4e1841b6c2b8de73d13e3e8db8)

In progress

**Comment by pandaring2you** [comment]
Jure raised issue found in test iteration 4 with Jon. Sifnode has a fix that Jure is testing

**Comment by pandaring2you** [comment]
Test now run, still need confirmation that the results are as expected. Now need to re-run with test with multiple nodes.

**Comment by pandaring2you** [comment]
Note from 7/22/2022 load testing meetingL Once this test is resolved. Will reconvene load testing meeting to examine results and determine additional load tests.

**Comment by jzvikart** [comment]
### Test iteration 7:
Parameters:
- number_of_liquidity_pools: 100
- number_of_wallets: 8000
- liquidity_providers_per_wallet: 3
- reward_period_default_multiplier: 1.0
- reward_period_distribute: True
- reward_period_mod: 1
- reward_period_pool_count: 100
- test_duration_blocks: 100
- number_of_nodes: 1

**Comment by sifag** [comment]
@jzvikart what are the errors reported on the 500s on block_results?

**Comment by bksifchain** [comment]
@jzvikart - Add nodes to scale up to the production set of ~125 nodes eventually and lets see the results

**Comment by jzvikart** [comment]
@sifag The errors look like this:
```
$ curl http://127.0.0.1:26657/block_results?block=
500 Internal Server Error
```

**Comment by jzvikart** [comment]
Open questions about multinode implementation:
- Does the topology of connections between the nodes matter?
- Does it matter which node should be used for sending transactions and which for querying?
- Does the network latency matter?

**Comment by daechoi** [comment]
1,  A<->B<->C<->A
2. Would like to know effect on response time for hitting different node query and txns.
Hit A for txn, A for query
Hit A for txn, B for query
Hit A for txn, C for query
3. Let’s put C on different region.

**Comment by jzvikart** [comment]
Due to time pressure we decided to ignore the effects of topology and latency for now. We will run the tests according to @gzukel's setup, and we will decide later if we need to do more exploration.

More information about multinode setup (Luca):
- https://docs.sifchain.finance/network-security/validators/running-sifnode-and-becoming-a-validator
- https://docs.sifchain.finance/developers/tutorials

**Comment by jzvikart** [comment]
@sifag found some additional information about HTTP 500 errors for `block_results`. The full response actually looks like this:

```
{
  "jsonrpc": "2.0",
  "id": -1,
  "error": {
    "code": -32603,
    "message": "Internal error",
    "data": "could not find results for height #883"
  }
```

The current hypothesis is that even though a new block is already being reported by `sifnoded status`, it might take some time for the data to become available through `get_results`.

To check this, we will modify the test to explicitly pass paratemer for `height` as `current_block - 1` instead of default (no `height` parameter aka current block). If the hypothesis is true, we should see a significantly reduced number of those errors.

**Comment by sifag** [comment]
Hypothesis is correct: I saw exactly zero 500s on my machine

**Comment by jzvikart** [comment]
There was a bug in previous versions of the test. Number of liquidity pools was fixed to 10. Because of this, the test might have been giving off wrong results, in particular the measured block time might have been lower than it should be.

**Comment by jzvikart** [comment]
### Test iteration 8:
Parameters:
- number_of_liquidity_pools: 100
- number_of_wallets: 8000
- liquidity_providers_per_wallet: 3
- reward_period_default_multiplier: 1.0
- reward_period_distribute: True
- reward_period_mod: 1
- reward_period_pool_count: 100
- test_duration_blocks: 100
- number_of_nodes: 1
- fix/lppd_rewards_block_time: yes
- fix for `--number-of-liquidity-pools`: yes

In progress

**Comment by jzvikart** [comment]
### Test iteration 9:

Summary: test code was almost completely rewritten to support multinode implementation, so we are running the new code with the same scenario as in iteration 8. We expect the results to be the same.

Pending

**Comment by jzvikart** [comment]
We found a bug in the rewards calculation code that halts the chain because of a broken assertion.

sifnoded.log: `ERR CONSENSUS FAILURE!!! err="negative coin amount" module=consensus`

We were able to reproduce the bug with the following parameters:

```
../integration/framework/venv/bin/python3 test_many_pools_and_liquidity_providers.py \
    --number-of-nodes 1 \
    --number-of-wallets 8000 \
    --number-of-liquidity-pools 100 \
    --liquidity-providers-per-wallet 3 \
    --reward-period-default-multiplier 1.0 \
    --reward-period-distribute \
    --reward-period-pool-count 100 \
    --reward-period-mod 1 \
    --lpd-period-mod 1 \
    --phase-offset-blocks 3 \
    --phase-duration-blocks 99999999 \
    --block-results-offset 1 \
    --run-forever | tee test_many_pools_and_liquidity_providers.log
```

Under these conditions, the error happens approx. 1.5h into from the start of the test.

**Comment by pandaring2you** [comment]
Note: as there have been several releases since this was last worked on, Jure will need to revisit adapting these tests with Pradeep and Caner. 

Primary goal is still to have load tests for LPD+Rewards and future features (PMTP updates, peggy2, etc.).

**Comment by pandaring2you** [comment]
Ex. Block time went up to 23 seconds, but currently having difficulty reproducing.

**Comment by pandaring2you** [comment]
Tests fixed and currently running - @jzvikart to check in on estimates of when tests will be fixed.

**Comment by jzvikart** [comment]
Test with 100 wallets and 3 LP/wallet failed to set up. The command "query clp lplist" is consistently timing out with `Error: post failed: Post "http://127.0.0.1:10286": EOF`.

Most likely cause is a large number of results.

**Comment by pandaring2you** [comment]
@jzvikart to file a ticket with sifnode for the issue above. This is the first time we are running a test with 100K wallets (300K LPs)

**Comment by pandaring2you** [comment]
On hold. Jure re-assigned to helping test margin.

**Comment by pandaring2you** [comment]
Note: since we're outsourcing to 3rd party providers, they might be more optimized. in that case, this issue is of less importance

**Comment by pandaring2you** [comment]
https://docs.google.com/spreadsheets/d/1hqO_VSk3VeaBwFe9XEdfSUEof1reQmG7qsg4EL7duP0/edit#gid=0

**Comment by pandaring2you** [comment]
On hold - even at 10X of current traffic levels, we should be fine on Peggy2. @jzvikart - can you re-run the load tests to double check they are working? Once this is done, we can close this task out (in the context of Peggy2). If we run into sifnode load issues - will create a separate issue.
